synapse/storage/databases/main/purge_events.py

```diff
@@ -319,14 +319,31 @@ async def purge_room(self, room_id: str) -> List[int]:
         Returns:
             The list of state groups to delete.
         """
-        return await self.db_pool.runInteraction(
+        state_groups_to_delete = await self.db_pool.runInteraction(
             "purge_room",
             self._purge_room_txn,
             room_id=room_id,
-            # This is safe because we don't care if room data is updated during the transaction
+            # This is safe because we don't care if room data is updated during the transaction, note
+            # we run a second transaction to cleanup tables we may write to during this transaction.
             isolation_level=IsolationLevel.READ_COMMITTED,
         )
 
+        def _purge_room_second_pass_txn(txn: LoggingTransaction, room_id: str) -> None:
+            for table in (
+                "event_push_actions",
+                "stream_ordering_to_exterm",
```

Comment:
Are these the only one's we care about? I think it might just be easier to re-run the same `_purge_room_txn` again a second time, as it should be quick if we've deleted the majority of stuff?

Reply:
🤦, so much simpler! https://github.com/matrix-org/synapse/pull/12942/commits/c83db929391d2dc35e7517f151741a1faac8070d